merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,872 @@
+
+        let offset = offset
+            + (mem::size_of::<Option<I>>()
+                * usize::try_from(hash_set_ref.capacity_indices).unwrap());
+        hash_set_ref.values =
+            NonNull::new(bytes.as_mut_ptr().add(offset) as *mut Option<HashSetCell>).unwrap();
+
+        hash_set_ref
+    }
+
+    // Insert a value into the hash set
+    pub fn insert(&mut self, value: &BigUint, sequence_number: usize) -> Result<(), HashSetError> {
+        for i in
+            0..usize::try_from(self.capacity_values).map_err(|_| HashSetError::IntegerOverflow)?
+        {
+            let i = I::try_from(i).map_err(|_| HashSetError::IntegerOverflow)?;
+            let probe_index = (value.clone() + i.to_biguint().unwrap() * i.to_biguint().unwrap())
+                % self.capacity_values.to_biguint().unwrap();
+            let probe_index = probe_index.to_usize().unwrap();
+            let index_bucket = unsafe { &mut *self.indices.as_ptr().add(probe_index) };
+
+            match index_bucket {
+                // The visited hash set cell points to a value in the array.
+                Some(value_index) => {
+                    let value_bucket = unsafe {
+                        &mut *self.values.as_ptr().add(
+                            usize::try_from(*value_index).map_err(|_| HashSetError::UsizeConv)?,
+                        )
+                    };
+                    match value_bucket {
+                        // The cell in the value array is already taken.
+                        Some(value_bucket) => {
+                            // We can overwrite that cell only if the element
+                            // is expired - when the difference between its
+                            // sequence number and provided sequence number is
+                            // greater than the threshold.
+                            if let Some(other_sequence_number) = value_bucket.sequence_number {
+                                let sequence_diff =
+                                    sequence_number.checked_sub(other_sequence_number);
+                                if let Some(sequence_diff) = sequence_diff {
+                                    if sequence_diff >= self.sequence_threshold {
+                                        *value_bucket = HashSetCell {
+                                            value: value.clone(),
+                                            sequence_number: None,
+                                        };
+                                        self.next_value_index = if self.next_value_index
+                                            < self.capacity_values - 1.into()
+                                        {
+                                            self.next_value_index + I::zero()
+                                        } else {
+                                            0.into()
+                                        };
```

Comment:
What's the purpose of this code?
(My intuition was that this is only being used when inserting new elements. Also self.next_index is not used in this match branch.)

Reply:
I removed it, it was useless